### ISSUE TITLE
New pipeline might break plugins by calling the register method twice

### ISSUE THREAD

**Issue by imriz** [elastic/logstash#4851]
Sample config:
```
input {

  udp {
    type => "syslog"
    buffer_size => 65536
    host => "0.0.0.0"
    port => 5140
    codec =>   plain {
      charset => "UTF-8"
    }
  }

}

output {
tcp {
    host => '127.0.0.1'
    port => '2048'
    mode => 'server'
}
}
```

Debug log:
```
{:timestamp=>"2016-03-20T16:24:48.723000+0200", :message=>"Starting tcp output listener", :address=>"127.0.0.1:2048", :level=>:info, :file=>"logstash/outputs/tcp.rb", :line=>"75", :method=>"register"}
{:timestamp=>"2016-03-20T16:24:48.742000+0200", :message=>"Will start workers for output", :worker_count=>1, :class=>LogStash::Outputs::Tcp, :level=>:debug, :file=>"logstash/output_delegator.rb", :line=>"34", :method=>"initialize"}
{:timestamp=>"2016-03-20T16:24:48.758000+0200", :message=>"Starting tcp output listener", :address=>"127.0.0.1:2048", :level=>:info, :file=>"logstash/outputs/tcp.rb", :line=>"75", :method=>"register"}
{:timestamp=>"2016-03-20T16:24:48.762000+0200", :message=>"Starting UDP listener", :address=>"0.0.0.0:5140", :level=>:info, :file=>"logstash/inputs/udp.rb", :line=>"60", :method=>"udp_listener"}
{:timestamp=>"2016-03-20T16:24:48.766000+0200", :message=>"The error reported is: \n  Address already in use - bind - Address already in use", :file=>"logstash/agent.rb", :line=>"222", :method=>"execute"}
{:timestamp=>"2016-03-20T16:24:48.775000+0200", :message=>["org/jruby/ext/socket/RubyTCPServer.java:118:in `initialize'", "org/jruby/RubyIO.java:853:in `new'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-tcp-2.0.2/lib/logstash/outputs/tcp.rb:76:in `register'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/output_delegator.rb:105:in `register'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/output_delegator.rb:105:in `register'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/pipeline.rb:170:in `start_workers'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/pipeline.rb:170:in `start_workers'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/pipeline.rb:123:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/agent.rb:211:in `execute'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/runner.rb:90:in `run'", "org/jruby/RubyProc.java:281:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.2-java/lib/logstash/runner.rb:95:in `run'", "org/jruby/RubyProc.java:281:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.22/lib/stud/task.rb:24:in `initialize'"], :file=>"logstash/agent.rb", :line=>"223", :method=>"execute"}
```

The reason is that the register method is being called twice -
Once in the https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/pipeline.rb#L58
and another in https://github.com/elastic/logstash/blob/592835cea239fa5afd943a4c77028d5821b706ff/logstash-core/lib/logstash/pipeline.rb#L207

This behaviour was introduced in 028576ba6f141f57c180c8f3dbffa14cb24967c0 (Issue #4340)

This might happen in several other plugins that do logic in the register method.

**Comment by ebuildy** [comment]
Strange I see only 1 call to input register:

https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/pipeline.rb#L373

And didn't have any error with new pipeline (released 2 months ago now).

**Comment by imriz** [comment]
@ebuildy I didn't look into input plugin, but there is a call in the eval here:
https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/pipeline.rb#L99

Can you reproduce the error with the above sample (simple) config?

**Comment by imriz** [comment]
To be more precise, the eval creates a new Logstash::OutputDelegator, which then calls the register in the initialize method.

**Comment by suyograo** [comment]
Looks like this will be solved in https://github.com/elastic/logstash/pull/4763

**Comment by imriz** [comment]
Yep, seems like it. Will it be released as part of 2.2.3?The way I see it, it is a major bug, as it makes at least some core plugins unusable, and in some can even result data loss (for the s3 plugin for example).

**Comment by suyograo** [comment]
@imriz absolutely. its going to be part of 2.2.3 and 2.3.0 releasing soon.

**Comment by imriz** [comment]
Thanks :)

**Comment by suyograo** [comment]
Fixed in  #4763
